src/tir/transforms/lower_async_dma.cc

```diff
@@ -192,19 +198,33 @@ class AsyncDMALowerer : public StmtExprMutator {
       // save queue ID for inspection in `wait` transform
       queue_ids_.insert(queue_id);
 
-      return Evaluate(Call(DataType::Int(32), builtin::dma_copy(),
-                           {queue_id,
-                            Call(DataType::Handle(), builtin::address_of(),
-                                 {BufferLoad(bufferstorenode->buffer, store_index)}),
-                            Call(DataType::Handle(), builtin::address_of(),
-                                 {BufferLoad(bufferloadnode->buffer, load_index)}),
-                            for_loop->extent * bufferloadnode->dtype.bytes(), dma_bypass_cache_}));
+      auto call_dma_copy =
+          Evaluate(Call(DataType::Int(32), builtin::dma_copy(),
+                        {queue_id,
+                         Call(DataType::Handle(), builtin::address_of(),
+                              {BufferLoad(bufferstorenode->buffer, store_index)}),
+                         Call(DataType::Handle(), builtin::address_of(),
+                              {BufferLoad(bufferloadnode->buffer, load_index)}),
+                         for_loop->extent * bufferloadnode->dtype.bytes(), dma_bypass_cache_}));
+
+      // if the buffer we are about to DMA was modified by the primfunc
+      // then we need to flush the buffer from the cache prior to the DMA
```

Comment:
This is great!  Nice general solution. Can we detect a buffer coming in the other direction?  In this direction it is a flush, coming back we need an invalidate.

Reply:
Agreed that it's probably better to perform an invalidation vs. flush depending on the directionality of the data transfer:

- Upon DMA “read”, you have to flush before

- Upon DMA “write”, you have to invalidate after

A helpful example hopefully could be what was done for VTA when VTA would be implemented with non-coherent DMA, as in here: https://github.com/apache/tvm/blob/bf0607bd317a7db8eba1a91c12170934c1ad201f/vta/runtime/runtime.cc#L1319-L1329

Reply:
@tmoreau89 and @janetsc I think this is really good feedback but I am a little leery to make changes without a failing unit test to use for test driven development as with `test_matmul.py` in this PR.  I imagine that software cache management to enable DMA bypass on Hexagon will be an iterative process.  It seems like you are pointing to the next iteration based on VTA example.  I would like to let this PR move through on its own merit and then address follow on cases, if possible.  Thoughts?

Reply:
So the next PR could have:

1.  Unit tests that expose the root of the problem (if they don't, we should go back to the drawing board)
2.  Additions to the matmul test that expose the need to invalidate in the VTCM to DDR direction
3. Changes to insert flush in the right place and invalidate in the right place.

The main reason I was advocating to do both directions now is that I'd like to confirm that those two unit tests do indeed fail.  If they don't, we'll need to do more experiments to get to the root of it.  The unit tests (from our offline discussion):

DDR to VTCM
    Allocate two buffers in DDR and VTCM
    Write "0xbeefbeef" to the DDR buffer.  Flush.
    Write "0xfaceface" to the DDR buffer.  Do NOT flush.
    DMA from DDR to VTCM with BYPASS ON.
Compare DDR and VTCM buffers.  They should not match, because the real values you wanted in DDR weren't flushed to DDR.

VTCM to DDR
    Allocate two buffers in DDR and VTCM
    Write "0xbeefbeef" to the DDR buffer.  (You can flush or not - it actually won't matter.)
    Write "0xfaceface" to the VTCM buffer.
    DMA from VTCM to DDR with BYPASS ON.
Compare DDR and VTCM buffers.  They will not match because there were stale values in the L2 cache for the DDR buffer, so it won't pick up what you wrote with BYPASS ON.  DDR needed to be invalidated first.

All that said, with a flush-only operation for DMA, I'll sign off on this as a good incremental step.  Thanks!